server/rss/queries.js

```diff
@@ -1,122 +1,257 @@
-					{
-						model: Release,
-						as: 'releases',
-					},
 				],
 			},
 		],
-	}).then((communityData) => {
-		if (!communityData) {
-			throw new Error('Community Not Found');
-		}
-		const communityUrl = communityData.domain
-			? `https://${communityData.domain}`
-			: `https://${communityData.subdomain}.pubpub.org`;
-		const feed = new RSS({
-			title: communityData.title,
-			description: communityData.description,
-			feed_url: `${communityUrl}/rss.xml`,
-			site_url: communityUrl,
-			image_url: communityData.favicon,
-			webMaster: 'hello@pubpub.org',
-			language: 'en',
-			pubDate: new Date(),
-			ttl: '60',
-		});
-		communityData.pubs
-			.map((pub) => {
-				const pubJSON = pub.toJSON();
-				const publishedDate = getPubPublishedDate(pub);
-				if (publishedDate) {
-					pubJSON.listedDate = publishedDate;
-				}
-				return pubJSON;
-			})
-			.filter((pub) => {
-				return pub.listedDate;
-			})
-			.sort((foo, bar) => {
-				if (foo.listedDate < bar.listedDate) {
-					return 1;
-				}
-				if (foo.listedDate > bar.listedDate) {
-					return -1;
-				}
-				return 0;
-			})
-			.forEach((pub) => {
-				const authors = pub.attributions
-					.map((attribution) => {
-						if (attribution.user) {
-							return attribution;
-						}
-						return ensureUserForAttribution(attribution);
-					})
-					.filter((item) => {
-						return item.isAuthor;
-					})
-					.map((item) => {
-						return item.user.fullName;
-					})
-					.join(', ');
+	});
+	// pubIds are already in publication-date order, so sort results based on that.
+	return pubs.sort((a, b) => pubIds.indexOf(a.id) - pubIds.indexOf(b.id));
+};
 
-				feed.item({
-					title: pub.title,
-					description: pub.description,
-					url: `${communityUrl}/pub/${pub.slug}`,
-					guid: pub.id,
-					author: authors,
-					date: pub.listedDate,
-					enclosure: {
-						url: pub.avatar,
-					},
-				});
-			});
+const createEnclosure = (url) => {
+	return {
+		enclosure: {
+			_attr: {
+				url: url,
+			},
+		},
+	};
+};
 
-		return feed.xml();
+export const getFeedItemForPub = (pubData, communityData) => {
+	const { title, description, collectionPubs, id } = pubData;
+	const formattedDownload = getFormattedDownloadUrl(pubData);
+	const pdfExport = getPublicExportUrl(pubData, 'pdf');
+	const xmlExport = getPublicExportUrl(pubData, 'jats');
+	const bestPdf = formattedDownload || pdfExport;
+	return {
+		title: title,
+		description: description,
+		url: pubUrl(communityData, pubData),
+		guid: id,
+		date: getPubPublishedDate(pubData),
+		categories: collectionPubs
+			.sort((a, b) => +b.isPrimary - +a.isPrimary)
+			.map(({ collection }) => collection.title),
+		custom_elements: [
+			...getAllPubContributors(pubData, false, true).map((attribution) => {
+				return {
+					'dc:creator': getContributorName(attribution),
+				};
+			}),
+			pubData.avatar && createEnclosure(pubData.avatar),
+			bestPdf && createEnclosure(bestPdf),
+			xmlExport && createEnclosure(xmlExport),
```

Comment:
Without reviewing the XML yet I assume the XML parser ignores `undefined` array elements instead of converting them to some weird or unexpected XML node?

Reply:
Yes it does, but I should probably filter here anyone so this is less confusing in the future!

Reply:
I just realized I used XML three times in the same sentence 😵

Reply:
![PlushUnrulyKoalabear-size_restricted](https://user-images.githubusercontent.com/2208769/92654739-47c4a880-f2be-11ea-8754-5e5644b2fa2b.gif)